### ISSUE TITLE
Contributing a high-level PAC function

### ISSUE THREAD

**Issue by choldgraf** [voytekresearch/pacpy#45]
Hey folks - I've got a PR right now in the MNE-sandbox repo, which adds a high-level function to do a bunch of PAC stuff between signals. It lets you:

1. Give an input that is shape (n_signals x time), and specify pairs of indices that you want to calculate PAC between.
2. Give a list of phase frequencies and amplitude frequencies to calculate PAC between.

It will then pre-filter the proper channels only one time per pair of low/high frequencies, so that we don't duplicate filtering etc.

You can also:

1. Give a list of indices corresponding to events
2. Give a list of tmin / tmax pairs either relative to event onsets, or to the whole dataset

It will then calculate PAC using the metric of choice for each pair of channel indices, phase/amp frequencies, and tmin/tmax window. With the option of either calculating this per epoch, or concatenating across epochs first.

So this way, you could make PAC plots with this function like this:

![image](https://cloud.githubusercontent.com/assets/1839645/15687469/d15db4de-2729-11e6-85dd-0e89947cc750.png)

Or like this:

![image](https://cloud.githubusercontent.com/assets/1839645/15687477/d97e28a6-2729-11e6-9327-0bf259d0bdd4.png)

Or do a connectivity plot w/ lots of pairs of electrodes.

In the PR we've been discussing whether it would make more sense to contribute this to PacPy, since there's a hidden function that's basically designed to not be specific for MNE objects, but instead operate on arrays. Would you all be interested in this? I'd be willing to take some time porting over the tests and the visualization functions (or whatever subset of that you're interested in). However I know that you guys generally want to keep this repo restricted to just the PAC metrics themselves.

LMKWYT. Pinging @srcole and @parenthetical-e in case you guys have thoughts ;)

**Comment by srcole** [comment]
I think it'd be great to add to pacpy! This new function would definitely be an improvement to the current 'comodulogram' function that currently repeats filtering and also doesn't have these nice visualizations.

**Comment by choldgraf** [comment]
The visualizations are pretty simple to implement, and I don't think they'd need to be part of a PR. The benefit of the function is that it can output PAC as an array of shape (n_pac_metrics, n_epochs, n_channel_pairs, n_frequency_pairs, n_time_windows) so that you can then slice and dice the output however you like :)

**Comment by srcole** [comment]
True. Yeah, I think this will be good to provide that functionality for people analyzing trial data sets with many simultaneous recordings.

**Comment by choldgraf** [comment]
@parenthetical-e what do you think? I believe you were the most hesitant to start adding extra functionality to pacpy (which I think has reasonable arguments behind it). IMO this is a close-enough extension to the functionality that's already there, so doesn't deviate too much from the overall goal of pacpy. What do you think?

**Comment by parenthetical-e** [comment]
I am all for adding the ability to work with trials and 2-d arrays of data, but am very much against the making a monolithic 

```python
phase_amplitude_coupling(..., method='plv')
```

Something like,

```python 
from pacpy.pac import plv
trials(lo, hi, f_lo, f_hi, trial_index, pacfn=plv)
channels(X f_lo, f_hi, lo_index, high_index, pacfn=plv) # where X is 2d
```

**Comment by parenthetical-e** [comment]
(Sorry. Had no intent to close this. I've thumb hands.)

**Comment by choldgraf** [comment]
@parenthetical-e I think I'm +1 to that, but let's chat about implementation etc. The reason that I've got all those flags is because I'm assuming that the input to the function is always the same type (channels x time) and it always does the filtering on that data first, before it does any slicing and dicing. That's why I don't have, e.g., an extra function for trials-shaped data. (and why this one is kind of a mammoth)

**Comment by choldgraf** [comment]
if you wanna skype and chat about the function etc, then I'm down for that. Only catch is that I'll be out of communication from Thursday to ~Tuesday (I'll be in Yellowstone hanging out with elks and stuff :D )

**Comment by parenthetical-e** [comment]
Let's just chat when you get back. Thursday the 9th? 10:30 AM?

**Comment by choldgraf** [comment]
wellllll, I might be leaving for knight lab camping on Thursday the 9th, so
wednesday would be better for me if that works for you.

**Comment by parenthetical-e** [comment]
Sure. 10 AM ok?

**Comment by choldgraf** [comment]
Sounds good to me, Wed June 8th, 10AM.

**Comment by srcole** [comment]
Works for me too

**Comment by parenthetical-e** [comment]
Cools.

**Comment by choldgraf** [comment]
Hey guys - I just realized that I had already opened this. Let's discuss integrating PAC efficiency etc into pacpy here.
